src/cinder/block_store.go

```diff
@@ -0,0 +1,259 @@
+	b.log.Infof("GetVolumeInfo called", volumeID, volumeAZ)
+
+	// Volume exists in list of volumes
+	if vol, ok := b.volumes[volumeID]; ok {
+		return vol.VolumeType, nil, nil
+	}
+
+	// Volume doesn't exist in list of volumes
+	return "", nil, fmt.Errorf("volume %v not found", volumeID)
+}
+
+// IsVolumeReady Check if the volume is ready.
+func (b *BlockStore) IsVolumeReady(volumeID, volumeAZ string) (ready bool, err error) {
+	b.log.Infof("IsVolumeReady called", volumeID, volumeAZ)
+
+	// If volume doesn't exist in list of volumes
+	_, ok := b.volumes[volumeID]
+	if !ok {
+		return false, fmt.Errorf("volume %v doesn't exist in volumes map", volumeID)
+	}
+
+	// Get volume object from Cinder
+	cinderVolume, err := volumes.Get(b.client, volumeID).Extract()
+	if err != nil {
+		b.log.Errorf("failed to get volume %v from Cinder", volumeID)
+		return false, err
+	}
+
+	// These are ready states
+	//   https://github.com/openstack/cinder/blob/master/api-ref/source/v3/volumes-v3-volumes.inc#volumes-volumes
+	if cinderVolume.Status == "available" || cinderVolume.Status == "in-use" {
+		return true, nil
+	}
+
+	// Volume is not in one of the "ready" states
+	return false, fmt.Errorf("volume %v is not in ready state, the status is %v", volumeID, cinderVolume.Status)
+}
+
+// CreateSnapshot creates a snapshot of the specified volume, and applies any provided
+// set of tags to the snapshot.
+func (b *BlockStore) CreateSnapshot(volumeID, volumeAZ string, tags map[string]string) (string, error) {
+	b.log.Infof("CreateSnapshot called", volumeID, volumeAZ, tags)
+	snapshotName := volumeID + ".snap." + strconv.FormatUint(rand.Uint64(), 10)
+
+	b.log.Infof("Trying to create snapshot", snapshotName)
+	opts := snapshots.CreateOpts{
+		Name:        snapshotName,
+		Description: "Velero snapshot",
+		Metadata:    tags,
+		VolumeID:    volumeID,
+	}
+
+	// Note: we will wait for snapshot to be in ready state in CreateVolumeForSnapshot()
+	createResult, err := snapshots.Create(b.client, opts).Extract()
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+	snapshotID := createResult.ID
+
+	// Remember the snapshot
+	b.snapshots[snapshotName] = snapshots.Snapshot{
+		Name:        snapshotName,
+		ID:          snapshotID,
+		Description: "Velero snapshot",
+		Metadata:    tags,
+		VolumeID:    volumeID,
+	}
+
+	b.log.Infof("Snapshot finished successfuly", snapshotName, snapshotID)
+	return snapshotName, nil
+}
+
+// DeleteSnapshot deletes the specified volume snapshot.
+func (b *BlockStore) DeleteSnapshot(snapshotName string) error {
+	b.log.Infof("DeleteSnapshot called", snapshotName)
+	snapshotID := b.snapshots[snapshotName].ID
+
+	// Delete snapshot from Cinder
+	b.log.Infof("Deleting Snapshot with ID", snapshotID)
+	err := snapshots.Delete(b.client, snapshotID).ExtractErr()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// Delete snapshot from list of snapshots
+	delete(b.snapshots, snapshotID)
```

Comment:
is snapshodID and snapshotName equal? Because you're using snapshotName key above.

Reply:
This part was deleted. Now I use only Snapshot IDs which are unique.